jina/optimize/__init__.py

```diff
@@ -0,0 +1,244 @@
+
+    def _load_env(self):
+        if self.env_yaml:
+            yaml = YAML(typ="safe")
+            self.env_parameters = yaml.load(open(self.env_yaml))
+            for environment_variable, value in self.env_parameters.items():
+                os.environ[environment_variable] = str(value)
+            logger.info("Environment variables loaded")
+        else:
+            logger.info("Cannot load environment variables as no env_yaml passed")
+
+    def _setup_workspace(self, workspace):
+        workspace.mkdir(exist_ok=True)
+        self.index_workspace = workspace / "index"
+        self.index_workspace.mkdir(exist_ok=True)
+        self.pod_workspace = workspace / "pods"
+        self.flow_workspace = workspace / "flows"
+        self.flow_workspace.mkdir(exist_ok=True)
+
+    def run_indexing(self, index_yaml, trial_parameters, workspace="workspace"):
+        self._load_env()
+        if workspace.exists():
+            if self.overwrite_workspace:
+                FlowRunner.clean_workdir(workspace)
+                logger.warning(
+                    colored("change overwrite_workspace to change this", "red")
+                )
+            else:
+
+                logger.warning(
+                    colored(
+                        "Workspace already exists. Skipping indexing.",
+                        "cyan",
+                    )
+                )
+                return
+
+        self._setup_workspace(workspace)
+        self._create_trial_pods(workspace, trial_parameters)
+        index_yaml = self._create_trial_flow(index_yaml, workspace)
+
+        self.index_document_generator, index_document_generator = tee(
+            self.index_document_generator
+        )
+
+        with Flow.load_config(index_yaml) as f:
+            f.index(index_document_generator, batch_size=self.index_batch_size)
+
+    def run_querying(
+        self, query_yaml, trial_parameters, callback, workspace="workspace"
+    ):
+        self._load_env()
+
+        self._setup_workspace(workspace)
+        self._create_trial_pods(workspace, trial_parameters)
+        query_yaml = self._create_trial_flow(query_yaml, workspace)
+
+        self.query_document_generator, query_document_generator = tee(
+            self.query_document_generator
+        )
+
+        with Flow.load_config(query_yaml) as f:
+            f.search(
+                query_document_generator,
+                batch_size=self.query_batch_size,
+                output_fn=callback,
+            )
+
+
+class EvaluationCallback:
+    def __init__(self, op_name=None):
+        self.op_name = op_name
+        self.evaluation_values = {}
+        self.n_docs = 0
+
+    def get_mean_evaluation(self):
+        if self.op_name:
+            return self.evaluation_values[self.op_name] / self.n_docs
+        return {
+            metric: val / self.n_docs for metric, val in self.evaluation_values.items()
+        }
+
+    def process_result(self, response):
```

Comment:
are we surr this evaluation will work on all documents? Isnt it only considering mean_value for only the first Request?

Reply:
The callback allows passing the eval name. The user can also write its own callback and pass to Optimizer.

Reply:
my point is that the callback only considers one response. but an optimizer cycle may consist of many Requests. 

So if I send 100 queries in batches of 10 it seems it will individually consider 10 at a time.

Reply:
It's giving mean evaluation of the first metric in the below dict (for all docs)
{metric: val / self.n_docs for metric, val in self.evaluation_values.items()}

Reply:
Yes, sorry the `=` led me to think it was not being updated